ruma-federation-api/src/transactions/send_transaction_message/edu.rs

```diff
@@ -0,0 +1,294 @@
+
+    #[doc(hidden)]
+    Custom(JsonValue),
+}
+
+#[derive(Debug, Deserialize)]
+struct EduTpeDeHelper {
+    /// The message type field
+    edu_type: String,
+    content: Box<RawJsonValue>,
+}
+
+impl<'de> de::Deserialize<'de> for Edu {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: de::Deserializer<'de>,
+    {
+        let json = Box::<RawJsonValue>::deserialize(deserializer)?;
+        let EduTpeDeHelper { edu_type, ref content } = from_raw_json_value(&json)?;
+
+        Ok(match edu_type.as_ref() {
+            "m.presence" => Self::Presence(from_raw_json_value(content)?),
+            "m.receipt" => Self::Receipt(from_raw_json_value(content)?),
+            "m.typing" => Self::Typing(from_raw_json_value(content)?),
+            "m.device_list_update" => Self::DeviceListUpdate(from_raw_json_value(content)?),
+            "m.direct_to_device" => Self::DirectToDevice(from_raw_json_value(content)?),
+            _ => Self::Custom(from_raw_json_value(content)?),
+        })
+    }
+}
+
+/// The content for "m.presence" Edu.
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub struct PresenceContent {
+    /// A list of presence updates that the receiving server is likely to be interested in.
+    pub push: Vec<PresenceUpdate>,
+}
+
+/// An update to the presence of a user.
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub struct PresenceUpdate {
+    /// The user ID this presence EDU is for.
+    pub user_id: UserId,
+
+    /// The presence of the user.
+    pub presence: PresenceState,
+
+    /// An optional description to accompany the presence.
+    #[serde(default, skip_serializing_if = "Option::is_none")]
+    pub status_msg: Option<String>,
+
+    /// The number of milliseconds that have elapsed since the user last did something.
+    pub last_active_ago: UInt,
+
+    /// Whether or not the user is currently active.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub currently_active: Option<bool>,
+}
+
+/// The content for "m.receipt" Edu.
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub struct ReceiptContent {
+    /// Receipts for a particular room.
+    #[serde(flatten)]
+    pub receipts: BTreeMap<RoomId, ReceiptMap>,
+}
+
+/// Mapping between user and `ReceiptData`.
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub struct ReceiptMap {
+    /// Read receipts for users in the room.
+    #[serde(rename = "m.read")]
+    pub read: BTreeMap<UserId, ReceiptData>,
+}
+
+/// Metadata about the event that was last read and when.
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub struct ReceiptData {
+    /// Metadata for the read receipt.
+    pub data: Receipt,
+
+    /// The extremity event ID the user has read up to.
+    pub event_ids: Vec<EventId>,
+}
+
+/// The content for "m.typing" Edu.
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub struct TypingContent {
+    /// The typing notification.
+    pub notifi: TypingNotice,
```

Comment:
The field is called `notifi`?

Reply:
It doesn't matter what it's called I need to add the `#[serde(flatten)]` attr to it should I just call it `notice` maybe?

Reply:
Sounds good.

Reply:
But why is the outer type needed at all then? Why isn't `TypingContent` the same as `TypingNotice`?

Reply:
Haha ugh! Yeah that was dumb I solved a problem I created for myself that wasn't there to begin with oops.